File: corpus/corpus.go
// Package corpus builds and reads the document set every engine is measured on.
//
// The corpus is one JSON lines file. That is the whole point of it. An earlier
// attempt measured each engine by pointing it at a directory tree, and the
// numbers that came back were mostly a measurement of the filesystem: a first
// pass over a checkout on Windows with a virus scanner in the way ran two orders
// of magnitude slower than a second pass over the same files, and no amount of
// staring at the engine explained the difference. Reading one sequential file
// that fits in the page cache removes that variable. What is left is the engine.
//
// It also makes the comparison fair in a way a directory does not. Engines
// disagree about which files are worth reading, how big a file has to be before
// it is skipped, and what counts as text. Deciding all of that once, here, means
// every engine is handed exactly the same documents in exactly the same order.
package corpus

import (
	"bufio"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"os"
	"path"
	"path/filepath"
	"strings"
	"unicode/utf8"
)

// MaxDocumentSize is the largest file taken into the corpus.
//
// Past a megabyte a checked in file is almost always generated, and a handful of
// them are large enough to dominate a total that is meant to describe a hundred
// thousand documents.
const MaxDocumentSize = 1 << 20

// Document is one record of the corpus.
//
// The fields are the ones every engine here can represent. Anything an engine
// cannot store is left out rather than approximated, so that no engine is
// charged for work another one is not doing.
type Document struct {
	// ID is unique across the corpus and stable between builds, so that two
	// runs on two machines are comparable.
	ID string `json:"id"`

	// Repo is the checkout the document came from, which is what makes a corpus
	// built from several projects still describable.
	Repo string `json:"repo"`

	// Path is the file path inside that checkout, using forward slashes.
	Path string `json:"path"`

	// Title is what a result list would show.
	Title string `json:"title"`

	// Body is the file contents.
	Body string `json:"body"`

	// Extension is the file extension without the dot, which is the one field
	// worth having as a filter because every engine can index it cheaply.
	Extension string `json:"ext"`
}

// Stats describes a corpus.
type Stats struct {
	Documents int   `json:"documents"`
	Bytes     int64 `json:"bytes"`
}

// Repo is one checkout to take into the corpus.
type Repo struct {
	// Name is what the repo field carries, and it is what a result refers to.
	// It is given rather than derived so that a corpus built from checkouts
	// scattered around a machine still names them the way people do.
	Name string

	// Dir is the checkout on disk.
	Dir string
}

// Write walks the checkouts under root and writes a corpus file.
//
// Each immediate subdirectory of root is treated as one project, and its name
// becomes the repo field. Nothing else about the layout matters.
func Write(root string, w io.Writer) (Stats, error) {
	entries, err := os.ReadDir(root)
	if err != nil {
		return Stats{}, err
	}

	var repos []Repo
	for _, e := range entries {
		if !e.IsDir() || skipDir(e.Name()) {
			continue
		}
		repos = append(repos, Repo{Name: e.Name(), Dir: filepath.Join(root, e.Name())})
	}
	if len(repos) == 0 {
		return Stats{}, fmt.Errorf("corpus: nothing to index under %s, is there a checkout in it", root)
	}
	return WriteRepos(repos, w)
}

// WriteRepos writes a corpus from checkouts that are named explicitly.
//
// This is the form worth using for a published result. A corpus that means the
// same thing on four machines has to be built from the same named projects at
// the same revisions, and that is easier to state as a list than as a directory
// that happens to contain the right things.
func WriteRepos(repos []Repo, w io.Writer) (Stats, error) {
	bw := bufio.NewWriterSize(w, 1<<20)
	enc := json.NewEncoder(bw)
	var st Stats

	for _, r := range repos {
		got, err := writeRepo(r.Dir, r.Name, enc)
		if err != nil {
			return st, err
		}
		st.Documents += got.Documents
		st.Bytes += got.Bytes
	}

	if st.Documents == 0 {
		return st, fmt.Errorf("corpus: no documents found in any of the %d checkouts given", len(repos))
	}
	return st, bw.Flush()
}

func writeRepo(dir, repo string, enc *json.Encoder) (Stats, error) {
	var st Stats
	err := filepath.WalkDir(dir, func(p string, d fs.DirEntry, err error) error {
		if err != nil {
			// A path that cannot be read is a fact about the checkout. Failing
			// the whole build over one of them would make the corpus depend on
			// which machine unpacked it.
			if d != nil && d.IsDir() {
				return fs.SkipDir
			}
			return nil
		}
		if d.IsDir() {
			if p != dir && skipDir(d.Name()) {
				return fs.SkipDir
			}
			return nil
		}
		if !d.Type().IsRegular() || !include(d.Name()) {
			return nil
		}
		info, err := d.Info()
		if err != nil || info.Size() > MaxDocumentSize {
			return nil
		}

		body, err := os.ReadFile(p)
		if err != nil {
			return nil
		}
		// A file that is not valid UTF-8 is a binary that happens to have a text
		// extension. Engines differ wildly in what they do with those, and none
		// of it is interesting.
		if !utf8.Valid(body) {
			return nil
		}

		rel, err := filepath.Rel(dir, p)
		if err != nil {
			return nil
		}
		rel = filepath.ToSlash(rel)

		doc := Document{
			ID:        repo + ":" + rel,
			Repo:      repo,
			Path:      rel,
			Title:     rel,
			Body:      string(body),
			Extension: strings.TrimPrefix(path.Ext(rel), "."),
		}
		if err := enc.Encode(doc); err != nil {
			return err
		}
		st.Documents++
		st.Bytes += int64(len(body))
		return nil
	})
	return st, err
}

// ErrStop is returned by the callback of [Read] to stop early without making
// it an error. It exists so that a runner asked for a fixed number of documents
// does not have to read the rest of a multi gigabyte file to find out it is
// done.
var ErrStop = errors.New("corpus: stop")

// Read calls fn for every document in a corpus file.
//
// It streams rather than returning a slice, because the corpus is larger than
// the heap most of these engines want to be measured with, and a runner that
// held the whole thing in memory before indexing it would be measuring
// something nobody deploys.
func Read(r io.Reader, fn func(Document) error) (Stats, error) {
	var st Stats
	sc := bufio.NewScanner(r)
	// Buffer for the largest line a corpus can hold: a document at the size
	// limit, plus the escaping JSON adds to it in the worst case.
	sc.Buffer(make([]byte, 0, 1<<20), 8*MaxDocumentSize)

	for sc.Scan() {
		line := sc.Bytes()
		if len(line) == 0 {
			continue
		}
		var d Document
		if err := json.Unmarshal(line, &d); err != nil {
			return st, fmt.Errorf("corpus: line %d: %w", st.Documents+1, err)
		}
		if err := fn(d); err != nil {
			if errors.Is(err, ErrStop) {
				return st, nil
			}
			return st, err
		}
		st.Documents++
		st.Bytes += int64(len(d.Body))
	}
	if err := sc.Err(); err != nil {
		return st, err
	}
	if st.Documents == 0 {
		return st, errors.New("corpus: file is empty")
	}
	return st, nil
}

// ReadFile is Read on a path.
func ReadFile(name string, fn func(Document) error) (Stats, error) {
	f, err := os.Open(name)
	if err != nil {
		return Stats{}, err
	}
	defer func() { _ = f.Close() }()
	return Read(f, fn)
}

// skipDir drops version control, dependency and build directories. They hold
// far more files than the content does and none of the prose.
func skipDir(name string) bool {
	switch name {
	case ".git", ".hg", ".svn", "node_modules", "vendor", "target", "dist", "build",
		".venv", "__pycache__", ".idea", ".vscode":
		return true
	}
	return strings.HasPrefix(name, ".")
}

// include is the list of extensions taken into the corpus. It is deliberately
// the same list the platform's directory connector uses, so that a number
// measured here describes the same work.
func include(name string) bool {
	if strings.HasPrefix(name, ".") {
		return false
	}
	switch name {
	case "OWNERS", "README", "LICENSE", "Makefile":
		return true
	}
	switch strings.ToLower(path.Ext(name)) {
	case ".md", ".markdown", ".rst", ".adoc", ".txt", ".html",
		".go", ".rs", ".py", ".js", ".ts", ".tsx", ".jsx", ".c", ".h", ".cc", ".cpp",
		".java", ".rb", ".sh", ".sql", ".yaml", ".yml", ".toml", ".json", ".proto":
		return true
	}
	return false
}

File: runners/rust/benchrs/src/corpus.rs
//! Reading the corpus file.
//!
//! It streams rather than collecting, because the corpus is larger than the
//! heap this engine wants to be measured with and a runner that held the whole
//! thing in memory before indexing it would be measuring something nobody
//! deploys.

use serde::Deserialize;
use std::fs::File;
use std::io::{BufRead, BufReader, Seek, SeekFrom};
use std::path::Path;

#[derive(Deserialize, Clone)]
pub struct Document {
    pub id: String,
    pub repo: String,
    pub path: String,
    pub title: String,
    pub body: String,
    pub ext: String,
}

/// Calls `f` for every document, and stops early when it returns false.
pub fn read<F>(name: &Path, mut f: F) -> std::io::Result<()>
where
    F: FnMut(Document) -> bool,
{
    let file = File::open(name)?;
    // A megabyte of buffer, because the documents are up to that size and a
    // smaller one turns a sequential read into a syscall per document.
    let reader = BufReader::with_capacity(1 << 20, file);
    for line in reader.lines() {
        let line = line?;
        if line.is_empty() {
            continue;
        }
        let doc: Document = serde_json::from_str(&line)
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;
        if !f(doc) {
            return Ok(());
        }
    }
    Ok(())
}

/// Splits the corpus into `count` byte ranges of roughly equal size.
///
/// The cuts land wherever the arithmetic puts them and not on a line boundary.
/// [`read_range`] is what makes them line up: a range starts at the first line
/// that begins after its start, and runs past its end to finish the line it is
/// in, so every document belongs to exactly one range and no document is split.
///
/// A count above the number of lines gives empty ranges rather than an error,
/// which is what a small corpus and a large machine should do.
pub fn shards(name: &Path, count: usize) -> std::io::Result<Vec<(u64, u64)>> {
    let len = std::fs::metadata(name)?.len();
    let count = count.max(1) as u64;
    Ok((0..count)
        .map(|i| (len * i / count, len * (i + 1) / count))
        .collect())
}

/// Calls `f` for every document whose line begins inside `[from, to)`.
///
/// This is how a runner reads the corpus on more than one thread. Each thread
/// opens the file itself and reads its own range, so the parsing goes as wide as
/// the indexing does. A runner that parsed on one thread and fanned out from
/// there would be measuring the parser.
pub fn read_range<F>(name: &Path, from: u64, to: u64, mut f: F) -> std::io::Result<()>
where
    F: FnMut(Document) -> bool,
{
    let mut file = File::open(name)?;
    let mut at = from.saturating_sub(u64::from(from > 0));
    file.seek(SeekFrom::Start(at))?;
    let mut reader = BufReader::with_capacity(1 << 20, file);
    let mut line = Vec::new();
    if from > 0 {
        // The line this range starts in the middle of belongs to the range
        // before it, which reads past its own end to finish it. Starting a byte
        // early is what makes a range that begins exactly on a line boundary
        // keep that line rather than let both ranges skip it.
        at += reader.read_until(b'\n', &mut line)? as u64;
    }
    while at < to {
        line.clear();
        let read = reader.read_until(b'\n', &mut line)?;
        if read == 0 {
            break;
        }
        at += read as u64;
        let text = line.strip_suffix(b"\n").unwrap_or(&line);
        if text.is_empty() {
            continue;
        }
        let doc: Document = serde_json::from_slice(text)
            .map_err(|e| std::io::Error::new(std::io::ErrorKind::InvalidData, e))?;
        if !f(doc) {
            return Ok(());
        }
    }
    Ok(())
}

/// Reads the query file, ignoring blank lines and comments.
pub fn queries(name: &Path) -> std::io::Result<Vec<String>> {
    let text = std::fs::read_to_string(name)?;
    Ok(text
        .lines()
        .map(str::trim)
        .filter(|l| !l.is_empty() && !l.starts_with('#'))
        .map(str::to_string)
        .collect())
}

#[cfg(test)]
mod tests {
    use super::*;

    /// The property every parallel runner rests on: the ranges together are the
    /// corpus, in order, with nothing read twice and nothing missed.
    #[test]
    fn the_shards_together_are_the_corpus() {
        let dir = std::env::temp_dir().join("benchrs-shards");
        std::fs::create_dir_all(&dir).expect("makes the directory");
        let path = dir.join("corpus.jsonl");
        let mut text = String::new();
        for i in 0..500 {
            text.push_str(&format!(
                "{{\"id\":\"{i}\",\"repo\":\"r\",\"path\":\"p\",\"title\":\"t\",\"body\":\"{}\",\"ext\":\"md\"}}\n",
                "word ".repeat(i % 40)
            ));
        }
        std::fs::write(&path, &text).expect("writes the corpus");

        let mut whole = Vec::new();
        read(&path, |d| {
            whole.push(d.id);
            true
        })
        .expect("reads");
        assert_eq!(whole.len(), 500);

        for count in [1, 2, 3, 7, 64, 700] {
            let mut parts = Vec::new();
            for (from, to) in shards(&path, count).expect("splits") {
                read_range(&path, from, to, |d| {
                    parts.push(d.id);
                    true
                })
                .expect("reads a range");
            }
            assert_eq!(parts, whole, "{count} shards");
        }
        std::fs::remove_dir_all(&dir).ok();
    }
}
